src/compiler/checker.ts

```diff
@@ -4092,7 +4106,15 @@ namespace ts {
                         const resolvedSymbol = resolveName(param, paramSymbol.name, SymbolFlags.Value, undefined, undefined);
                         paramSymbol = resolvedSymbol;
                     }
-                    parameters.push(paramSymbol);
+                    if (paramSymbol.name === "this") {

```

Comment:
Parser allows `this` parameters in any position, so here I would change to `if (i === firstParamIndex && paramSymbol.name === "this")` and then add `const firstParamIndex = isJSConstructSignature ? 1: 0` above the for loop.